fem/gridfunc.cpp

```diff
@@ -1338,107 +1338,253 @@ void GridFunction::GetVectorGradientHat(
    MultAtB(loc_data_mat, dshape, gh);
 }
 
-double GridFunction::GetDivergence(ElementTransformation &tr) const
+double GridFunction::GetDivergence(ElementTransformation &T) const
 {
-   double div_v;
-   int elNo = tr.ElementNo;
-   const FiniteElement *FElem = fes->GetFE(elNo);
-   if (FElem->GetRangeType() == FiniteElement::SCALAR)
+   switch (T.ElementType)
    {
-      MFEM_ASSERT(FElem->GetMapType() == FiniteElement::VALUE,
-                  "invalid FE map type");
-      DenseMatrix grad_hat;
-      GetVectorGradientHat(tr, grad_hat);
-      const DenseMatrix &Jinv = tr.InverseJacobian();
-      div_v = 0.0;
-      for (int i = 0; i < Jinv.Width(); i++)
+      case ElementTransformation::ELEMENT:
       {
-         for (int j = 0; j < Jinv.Height(); j++)
+         int elNo = T.ElementNo;
+         const FiniteElement *fe = fes->GetFE(elNo);
+         if (fe->GetRangeType() == FiniteElement::SCALAR)
+         {
+            MFEM_ASSERT(fe->GetMapType() == FiniteElement::VALUE,
+                        "invalid FE map type");
+            DenseMatrix grad_hat;
+            GetVectorGradientHat(T, grad_hat);
+            const DenseMatrix &Jinv = T.InverseJacobian();
+            double div_v = 0.0;
+            for (int i = 0; i < Jinv.Width(); i++)
+            {
+               for (int j = 0; j < Jinv.Height(); j++)
+               {
+                  div_v += grad_hat(i, j) * Jinv(j, i);
+               }
+            }
+            return div_v;
+         }
+         else
          {
-            div_v += grad_hat(i, j) * Jinv(j, i);
+            // Assuming RT-type space
+            Array<int> dofs;
+            fes->GetElementDofs(elNo, dofs);
+            Vector loc_data, divshape(fe->GetDof());
+            GetSubVector(dofs, loc_data);
+            fe->CalcDivShape(T.GetIntPoint(), divshape);
+            return (loc_data * divshape) / T.Weight();
          }
       }
+      break;
+      case ElementTransformation::BDR_ELEMENT:
+      {
+         // In order to capture the derivative of the normal component of
+         // the field we must evaluate it in the neighboring element.
+         FaceElementTransformations * FET =
+            fes->GetMesh()->GetBdrFaceTransformations(T.ElementNo);
+
+         // Boundary elements and Boundary Faces may have different
+         // orientations so adjust the integration point if necessary.
+         int o = 0;
+         if (fes->GetMesh()->Dimension() == 3)
+         {
+            int f;
+            fes->GetMesh()->GetBdrElementFace(T.ElementNo, &f, &o);
+         }
+
+         IntegrationPoint fip;
+         be_to_bfe(FET->GetGeometryType(), o, T.GetIntPoint(), fip);
+
+         FET->SetIntPoint(&fip);
+         ElementTransformation & T1 = FET->GetElement1Transformation();
+
+         return GetDivergence(T1);
```

Comment:
This returns the full divergence.  The derivative in the normal direction depends on all of the element's DoFs not just those on the face where as the tangential derivatives could be calculated using only face DoFs.  These comments were only meant to highlight why we go the extra trouble of evaluating these derivatives in the neighboring element.  They should have been more clear.  Thanks!

Reply:
But actually that is maybe what you want calculate, because it is also mathematically incorrect to evaluate it inside. For a derivative, you need a neighborhood of the point, which you do not have at the boundary of an element. This is especially counterintuitive for discontinuous functions, where you solve it by taking the derivative only on the first side, but why? it could be the other one, or average of both, etc.
To summarize, I am not sure that the generalization is that straightforward. Basically, there are two approaches to it, which I see:
- reduction to the lower dimension of the boundary/face element through Stokes' theorem. In the case of the divergence, it would the normal component for the boundary elements and the normal jump for internal faces. This is what came to my mind at first when reading about the generalization.
- evaluation of the derivative on the inner volume, but the user should have freedom in choice of the behavior or it should be somehow prevalent option. However, I do not see how the inner volume divergence on the first side of the face is useful practically.

Reply:
This method is only implemented for element interiors and the mesh boundary so there is no other side of the face to be concerned with.

I agree that computing the divergence on the boundary is odd and may not be useful in most cases.  I would prefer to keep this method as it is but offer two other alternatives that calculate either `du/dn` or `\nabla_t u`.

Reply:
I do not know, what do others think? @vladotomov , @psocratis , @jakubcerveny , should be the generalization of the differential operators just their mere evaluation on the inner volume of the element OR reduction to the lower dimension by means of the Stokes' theorem, i.e. normal jump for divergence and tangential jump for curl? I think the latter might be useful for someone and can be implemented also for the inner faces. This is something different than the generalization of GetValue() in #1194 , where point-wise evaluation at the boundary is the only meaningful thing you may do (except DGs, but the trace is what you normally mean).

Reply:
@najlkin I think that should probably be a different PR with different function names.